alibi_detect/saving/schemas.py

```diff
@@ -52,19 +49,22 @@ def coerce_int2list(value: int) -> List[int]:
         return value
 
 
-def validate_model(model: Callable, values: dict) -> Callable:
-    """Validator to check the model is compatible with the given backend"""
-    backend = values['backend']
-    if backend == 'tensorflow' and not isinstance(model, SupportedModels_tf):
-        raise ValueError('A TensorFlow backend is not available for this model')
-    elif backend == 'pytorch' and not isinstance(model, SupportedModels_torch):
-        raise ValueError('A PyTorch backend is not available for this model')
-    elif backend == 'sklearn' and not isinstance(model, SupportedModels_sklearn):
-        raise ValueError('A sklearn backend is not available for this model')
-    return model
+class SupportedModelsType:
+    """
```

Comment:
Previously we typed `model` fields as:

```python
model: Optional[Callable] = None
```

and then validated them with:

```python
_validate_model = validator('model', allow_reuse=True, pre=True)(validate_model)
```

This did not work for sklearn models since `sklearn.base.BaseEstimator` is not a `Callable`. We could simple change to `model: Optional[Any] = None`. However I think a more elegent solution is to use a pydantic [custom data type](https://pydantic-docs.helpmanual.io/usage/types/#custom-data-types). This can then be used directly as:

```python
model: Optional[SupportedModelsType] = None
```

Note, that as of now the validation we are doing has changed. We are simply validating if the `model` type is one of the `SupportedModels`, which itself is conditional on the installed optional deps. We could add in more specific checks to validate `model` against `SupportModels_tf` etc depending on `backend`. (just thought if this and will look into now!)

Reply:
Note that the:

```python
raise ValueError('A TensorFlow backend is not available for this model')
```

Has been changed to:

```python
raise TypeError("`backend='tensorflow'` but the `model` doesn't appear to be a TensorFlow supported model.")
```

as I think the previous error message was misleading.